### PR TITLE
Delay GPU->CPU sync in sampling

### PR DESCRIPTION
**Pull request by Yard1**
This PR preallocates tensors used to prune hidden states and to index the samples by sampling type allowing us to delay the GPU->CPU sync slightly longer (up until `torch.multinomial` in the sampler).

Should fix the slight performance regression introduced in https://github.com/vllm-project/vllm/pull/1309

Technically, we can delay the sync even further by reordering some more operations, but that's left for a future PR.

### CODE REVIEW

**Comment by Yard1** [comment]
cc @zhuohan123

**Comment by Yard1** [comment]
updating the PR!

**Comment by hanzhi713** [comment]
This would be a nice improvement to have. Since you're adding more h2ds, a small optimization would be to batch the h2ds together and make it async. 

image

**Comment by Yard1** [comment]
Yes, we could certainly optimize that as well - I would like to do that in a followup.

**Comment by Yard1** [comment]
@zhuohan123 Updated, PTAL!

File: vllm/worker/worker.py [hunk 1]
```diff
@@ -180,6 +184,24 @@ def _prepare_inputs(
...
             prompt_lens.append(prompt_len)
 
+            assert len(seq_ids) == 1, "Prompt input should have only one seq."
+            if sampling_params.prompt_logprobs is not None:
+                selected_token_indices.extend(
+                    range(selected_token_start_idx,
+                          selected_token_start_idx + prompt_len - 1))
+            selected_token_indices.append(selected_token_start_idx +
+                                          prompt_len - 1)
+            selected_token_start_idx += prompt_len
```
Comment (zhuohan123) [inline vllm/worker/worker.py:194]: To adapt to the changes in #1381:
```suggestion
            selected_token_start_idx += input_metadata.max_prompt_len
```

File: vllm/worker/worker.py [hunk 2]
```diff
@@ -180,6 +184,24 @@ def _prepare_inputs(
...
+
+            if sampling_params.prompt_logprobs is not None:
+                # NOTE: prompt token positions do not need sample, skip
+                categorized_sample_indices_start_idx += prompt_len - 1
+
+            categorized_sample_indices[sampling_params.sampling_type].extend(
+                range(categorized_sample_indices_start_idx,
+                      categorized_sample_indices_start_idx + 1))
```
Comment (zhuohan123) [inline vllm/worker/worker.py:202]: Nit:
```suggestion
            categorized_sample_indices[sampling_params.sampling_type].append(
                categorized_sample_indices_start_idx)
```

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thank you for your contribution! Left some small comments based on recent PRs. Let's refactor the bloated `input_metadata` in a future PR.

**Comment by Yard1** [comment]
Need to fix max_prompt_len in worker

**Comment by hanzhi713** [comment]
Any progress on this PR?

**Comment by Yard1** [comment]
@hanzhi713 I am on holidays currently but will wrap it up next week!

**Comment by Yard1** [comment]
Should be good now

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thank you for your contribution!

**Comment by gesanqiu** [comment]
@Yard1 @zhuohan123 Have you been test the sampling performance on AWQ model? I found that the AWQ model's first sampling process is much slower than FP16 model, but faster in the following tokens.

**Comment by zhaoyang-star** [comment]
@Yard1 Sorry for that I am not familar with `selected_token_indices`.

Assume we have 3 prompts. The length are 9, 8, 6. On the prompt (prefill) stage, selected_token_indices=[5, 14, 23]. But I think that selected_token_indices should be [8, 16, 22]. Below is a diagram, which is relatively more intuitive. I have been struggling to understand it here, and I would appreciate your interpretation.
selected_token_ids

**Pull request merged**